appengine/src/main/java/com/google/googleinterns/gscribe/resources/AuthenticationResource.java

```diff
@@ -0,0 +1,135 @@
+import com.google.googleinterns.gscribe.resources.io.exception.UserNotAuthorizedException;
+import com.google.googleinterns.gscribe.resources.io.request.AuthenticationRequest;
+import com.google.googleinterns.gscribe.resources.io.response.AuthenticationResponse;
+import com.google.googleinterns.gscribe.services.TokenService;
+import com.google.googleinterns.gscribe.services.data.TokenResponse;
+import com.google.inject.Inject;
+
+import javax.validation.constraints.NotNull;
+import javax.ws.rs.*;
+import javax.ws.rs.core.MediaType;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+
+@Path("/authenticate")
+public class AuthenticationResource {
+
+    private final TokenService tokenService;
+    private final UserTokenDao userTokenDao;
+
+    @Inject
+    public AuthenticationResource(TokenService tokenService, UserTokenDao userTokenDao) {
+        this.tokenService = tokenService;
+        this.userTokenDao = userTokenDao;
+    }
+
+    /**
+     * Get corresponding userID from the IDToken using tokenVerifier
+     * With the help of tokenGenerator generate the tokens from authCode
+     * Validate that userID from authCode is same as userID from IDToken in header
+     * Save the tokens in the database
+     *
+     * @param IDToken ( from header )
+     * @param request ( must contain authCode )
+     * @return a response message if tokens are saved in database for the user
+     * @throws MissingRequestParametersException    ( if request does not have authCode )
+     * @throws InvalidIDTokenException              ( if IDToken is invalid )
+     * @throws InternalServerErrorException         ( by GeneralSecurityException and IOException for credentials file )
+     * @throws InvalidAuthorizationRequestException ( if authCode is invalid, or user corresponding to IDToken and authCode are different )
+     */
+    @POST
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes(MediaType.APPLICATION_JSON)
+    public AuthenticationResponse saveToken(@NotNull @HeaderParam("Authentication") String IDToken, @NotNull AuthenticationRequest request) {
+        UserToken token;
+        TokenResponse tokenResponse;
+        String userID, tokenResponseUserID;
+
+        if (request.getAuthCode() == null)
+            throw new MissingRequestParametersException();
+
+        try {
+            userID = tokenService.verifyIDToken(IDToken);
+        } catch (GeneralSecurityException | IOException e) {
+            throw new InternalServerErrorException();
+        } catch (IllegalArgumentException e) {
+            throw new InvalidIDTokenException();
+        }
+
+        try {
+            tokenResponse = tokenService.generateToken(request.getAuthCode());
+        } catch (GeneralSecurityException | IOException e) {
+            throw new InvalidAuthorizationRequestException();
+        }
+
+        try {
+            tokenResponseUserID = tokenService.verifyIDToken(tokenResponse.getIDToken());
```

Comment:
We have already verified IdToken why are we verifying again?

Reply:
this is the IDToken derived from the authorization code. This will give us the userID for which this authorization code was meant for and we will check with original user request user ID that they match.